### ISSUE TITLE
lightning: Find a better way to handle `next_step`.

### ISSUE THREAD

**Issue by daavoo** [iterative/dvclive#477]
>  It was the DeviceStatsMonitor callback from Lightning! It tried to log every single validation step and really slowed the whole thing down.

We still need to figure out a better way of deciding when to call `next_step` in the lightning integration.

In other frameworks, the callbacks for integrations have a clear interface with events like `on_epoch_end`, but lightning Logger is a little messy because you only have access to `log_metrics`, and this can be called by other callbacks and at different paces.

**Comment by dberenbaum** [comment]
tldr: Let's call `next_step` for train metrics but not for val metrics. A lot of what's below is for reference so I can remember how lightning logging works 😄 .

To clarify what's happening here, let's assume https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#log is called with `on_step=True` and `on_epoch=True`. This is what will happen:

1. Lightning first iterates over all train batches for an epoch. By default, it will log every 50 train steps/batches.
2. Only after all train batches for an epoch have been run, then all val batches are run. By default, it will log every 1 step/batch.
3. After all train and val batches have been run, then val epoch metrics are logged (but the step is now the train step and not the val step), and finally the train epoch metrics are logged.

Example:

* train size: 800
* val size: 200
* batch size: 10
* epochs: 2

Lightning will log metrics with the following values:

```
step: 49, metrics: train_loss_step...
step: 99, metrics: train_loss_step...
...
step: 799, metrics: train_loss_step...
step: 0, metrics: val_loss_step...
step: 1, metrics: val_loss_step...
step: 2, metrics: val_loss_step...
...
step: 19, metrics: val_loss_step...
step: 999, metrics: val_loss_epoch...
step: 999, metrics: train_loss_epoch...
step: 1049, metrics: train_loss_step...
step: 1049, metrics: train_loss_step...
...
step: 1999, metrics: train_loss_step...
step: 1999, metrics: train_loss_step...
step: 20, metrics: val_loss_step...
step: 21, metrics: val_loss_step...
...
step: 39, metrics: val_loss_step...
step: 1999, metrics: train_loss_step...
step: 1999, metrics: train_loss_step...
```

This all very flexible and configurable, but the default setup means no val metrics will be recorded for a long time, followed by a lot of val metrics, each with their own step count unrelated to the train steps.

We could call `next_step` when we see the `epoch` metric get updated, which would get us close to what we do with other frameworks like keras where we currently only log at the end of each epoch. I think it would be better to try to respect the logging configuration so that, for example, updates can get shown more than once per epoch. We could instead try to call `next_step` for train metrics but not val metrics. This seems closer to the lightning philosophy, where `Trainer.log_every_n_steps` controls when logging happens but is completely separate from validation logging, which is instead controlled by `Trainer.val_check_interval` and `Trainer.check_val_every_n_epoch` (see https://pytorch-lightning.readthedocs.io/en/stable/common/trainer.html#trainer-class-api).

**Comment by dberenbaum** [comment]
> We could instead try to call `next_step` for train metrics but not val metrics.

I missed that there's no way to differentiate between them except by however the user decides to name them.

> We could call `next_step` when we see the `epoch` metric get updated

I think this is the best we can do.
